build(docs-infra): add `final` labels to classes in API pages

petebacondarwin opened the pull request:
This change will mark classes as `final` unless they have been annotated
with an `@extensible` jsdoc tag.

Fixes #42802

For example:

<img width="565" alt="Screenshot 2021-07-09 at 16 50 06" src="https://user-images.githubusercontent.com/15655/125104864-bf798480-e0d5-11eb-8aa8-7051585a1137.png">

petebacondarwin committed:
build(docs-infra): add `sealed` markers to classes in API pages

This change will mark classes as `sealed` unless they have been annotated
with an `@extensible` jsdoc tag.

Fixes #42802

mary-poppins commented:
You can preview b4564e0 at https://pr42807-b4564e0.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/templates/api/base.template.html:
```diff
@@ -26,6 +26,7 @@ <h1>{$ doc.name | escape $}</h1>
     {% if doc.deprecated !== undefined %}<label class="api-status-label deprecated">deprecated</label>{% endif %}
     {% if doc.security !== undefined %}<label class="api-status-label security">security</label>{% endif %}
     {% if doc.pipeOptions.pure === 'false' %}<label class="api-status-label impure-pipe">impure</label>{% endif %}
+    {% if doc.docType === 'class' and doc.extensible !== true %}<label class="api-status-label sealed" title="This class should not be extended.">sealed</label>{% endif %}
```
Since "seal"/"sealed" have a [slightly different meaning][1] in JS-land, I suggest using another term to avoid confusion.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/seal

petebacondarwin replied:
Suggestions?

petebacondarwin replied:
`final` is too vague IMO.
`inextensible` or `do not extend` are too verbose.

gkalpak replied:
`final` and `not extensible` were my only two ideas, hence I didn't offer any suggestions :grin:
And there's no relevant feature/term in JS/TS afaict 🤷‍♂️

AndrewKushnir replied:
Can we may be make the label clickable and forward users to a new (special) section of the PUBLIC_API.md file that would contain more info? If yes, I'd vote for the "final" as a label.

petebacondarwin replied:
Making it a link is easy enough. I'll add that and change it to `final`.

mary-poppins commented:
You can preview 755fdf2 at https://pr42807-755fdf2.ngbuilds.io/.

petebacondarwin commented:
@gkalpak and @AndrewKushnir -  I changed the label to `final` and made it a link. I am not sure this is a good UI though (to have one of these labels clickable). So I also tried out adding a line of boilerplate text under the short-description too.

For an example see https://pr42807-755fdf2.ngbuilds.io/api/forms/FormGroup

<img width="583" alt="Screenshot 2021-07-12 at 12 28 46" src="https://user-images.githubusercontent.com/15655/125280435-f2f22400-e30c-11eb-8482-93b58162d95f.png">


Please take a look and let me know your thoughts.

petebacondarwin commented:
OK, third iteration: George and I thought that putting the "This class is "final"..." text in the short description drew too much attention to it, and would actually be more confusing.

So now, I have moved that text to the constructor's short description (if there is a constructor).

<img width="864" alt="Screenshot 2021-07-12 at 16 25 03" src="https://user-images.githubusercontent.com/15655/125314896-87b94980-e32e-11eb-91a0-7cfc15fd4497.png">

mary-poppins commented:
You can preview 1717ccb at https://pr42807-1717ccb.ngbuilds.io/.

gkalpak commented on aio/src/styles/2-modules/api-pages/_api-pages.scss:
```diff
@@ -97,7 +97,7 @@
         .github-links {
           a {
-
+            
```
🤨

petebacondarwin replied:
Doh!

gkalpak commented on aio/tools/transforms/templates/api/lib/memberHelpers.html:
```diff
@@ -95,6 +95,14 @@ <h3>
+
+        {% if method.containerDoc.docType === 'class' and method.name === 'constructor' and method.containerDoc.extensible !== true %}
+        <p class="final-message">
+          This class is "final" should not be extended.
```
I feel like something is missing between the words "final" and "should". Maybe _"is final **and** should not"_?

petebacondarwin replied:
Indeed!

gkalpak commented on aio/tools/transforms/templates/api/base.template.html:
```diff
@@ -26,6 +26,9 @@ <h1>{$ doc.name | escape $}</h1>
     {% if doc.security !== undefined %}<label class="api-status-label security">security</label>{% endif %}
     {% if doc.pipeOptions.pure === 'false' %}<label class="api-status-label impure-pipe">impure</label>{% endif %}
+    {% if doc.docType === 'class' and doc.extensible !== true %}<label class="github-links api-status-label final" title="This class should not be extended.">
+      <a href="https://github.com/angular/angular/blob/master/docs/PUBLIC_API.md#final-classes">final</a>
```
I am a little torn on using `master` vs using a specific commit SHA. Leaving it up to you (but wanted to bring it up to make sure it has been considered).

petebacondarwin replied:
I think we can actually reuse the logic used in the github link anchors. Let me look into that.

mary-poppins commented:
You can preview 74291f7 at https://pr42807-74291f7.ngbuilds.io/.

mary-poppins commented:
You can preview aef35d3 at https://pr42807-aef35d3.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Looks great, thanks Pete 👍

petebacondarwin committed:
docs(upgrade): add `@extensible` to the `UpgradeComponent` class

This class is designed to be used as a base class. Therefore it should not be
marked as `final` in the API docs.

petebacondarwin commented:
I realised that we cannot merge this until we have ensure that all extensible classes are not marked as final.

I updated the logic so that abstract classes are never marked as `final`.
Apart from that the only concrete class I could find that is designed to be extended is `UpgradeComponent`.
@gkalpak / @AndrewKushnir do you know of any others?

mary-poppins commented:
You can preview 4e253db at https://pr42807-4e253db.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/templates/api/base.template.html:
```diff
@@ -26,6 +26,9 @@ <h1>{$ doc.name | escape $}</h1>
     {% if doc.deprecated !== undefined %}<label class="api-status-label deprecated">deprecated</label>{% endif %}
     {% if doc.security !== undefined %}<label class="api-status-label security">security</label>{% endif %}
     {% if doc.pipeOptions.pure === 'false' %}<label class="api-status-label impure-pipe">impure</label>{% endif %}
+    {% if doc.docType === 'class' and doc.extensible !== true and not doc.isAbstract %}<label class="github-links api-status-label final" title="This class should not be extended.">
```
What about other `docType`s, such as `ngmodule` (see for example [CommonModule][1]) or `directive` (see for example [NgIf][2])?

Don't we need to also mark _them_ as "final" (even if they don't have a public constructor most of the time)?

##
(NOTE: If we change the logic here, the same changes should be applied in [memberHelpers.html][3].)

[1]: https://pr42807-4e253db.ngbuilds.io/api/common/CommonModule
[2]: https://pr42807-4e253db.ngbuilds.io/api/common/NgIf
[3]: https://github.com/angular/angular/blob/4e253dbaab3ba5bd4f6cab53dbaa80fb658cbfd6/aio/tools/transforms/templates/api/lib/memberHelpers.html#L99

petebacondarwin replied:
Yeah - I decided that we should start with only classes, since they are more likely to feel like they can be extended. Also those kinds of classes are rendered in the docs in a way that doesn't make it so obvious that they are classes.

We can easily add that at a later stage if it looks useful.

gkalpak reviewed (approved):
One question. Otherwise lgtm 🚀

petebacondarwin commented:
@AndrewKushnir - can you override the presubmit check, since this PR only changes a comment in code that will be synched to G3?

mary-poppins commented:
You can preview 5879ed9 at https://pr42807-5879ed9.ngbuilds.io/.

AndrewKushnir commented:
> can you override the presubmit check, since this PR only changes a comment in code that will be synched to G3?

@petebacondarwin done! I'm adding "merge-assistance" label and this PR should be ready for merge. Thanks for implementing that Pete 👍

AndrewKushnir commented:
Merge-assistance: Pete has global docs approval permissions and this PR was reviewed and approved by 2 engineers, so this PR is ready for merge now.
